components/hab/src/command/pkg/download.rs

```diff
@@ -0,0 +1,439 @@
+                           -> Result<()>
+        where T: UIWriter
+    {
+        ui.status(Status::Downloading,
+                  format!("{} public origin key", &name_with_rev))?;
+        let (name, rev) = parse_name_with_rev(&name_with_rev)?;
+        self.api_client.fetch_origin_key(&name,
+                                          &rev,
+                                          token,
+                                          &self.path_for_keys(),
+                                          ui.progress())?;
+        ui.status(Status::Cached,
+                  format!("{} public origin key", &name_with_rev))?;
+        Ok(())
+    }
+
+    fn verify_artifact<T>(&self,
+                          ui: &mut T,
+                          ident: &PackageIdent,
+                          target: PackageTarget,
+                          artifact: &mut PackageArchive)
+                          -> Result<()>
+        where T: UIWriter
+    {
+        // We need to look at the artifact to know the signing keys to fetch
+        // Once we have them, it's the natural time to verify.
+        // Otherwise, it might make sense to take this fetch out of the verification code.
+        let nwr = artifact::artifact_signer(&artifact.path)?;
+        if SigKeyPair::get_public_key_path(&nwr, &self.path_for_keys()).is_err() {
+            ui.status(Status::Downloading,
+                      format!("Public key for signer {:?}", nwr))?;
+            self.fetch_origin_key(ui, &nwr, self.token)?;
+        }
+
+        if self.verify {
+            ui.status(Status::Verifying, artifact.ident()?)?;
+            artifact.verify(&self.path_for_keys())?;
+            debug!("Verified {} for {} signed by {}", ident, target, &nwr);
+        }
+        Ok(())
+    }
+
+    // This function and it's sibling in install.rs deserve to be refactored to eke out commonality.
+    fn is_artifact_downloaded(&self, ident: &PackageIdent, target: PackageTarget) -> bool {
+        self.downloaded_artifact_path(ident, target).is_file()
+    }
+
+    // This function and it's sibling in install.rs deserve to be refactored to eke out commonality.
+    /// Returns the path to the location this package would exist at in
+    /// the local package cache. It does not mean that the package is
+    /// actually *in* the package download directory, though.
+    fn downloaded_artifact_path(&self, ident: &PackageIdent, target: PackageTarget) -> PathBuf {
+        self.path_for_artifact()
+            .join(ident.archive_name_with_target(target).unwrap())
+    }
+
+    fn fetch_latest_package_in_channel_for(&self,
+                                           ident: &PackageIdent,
+                                           target: PackageTarget,
+                                           channel: &ChannelIdent,
+                                           token: Option<&str>)
+                                           -> Result<Package> {
+        let origin_package = self.api_client
```

Comment:
Minor - you can eliminate the let and Ok()... just return show_package_metadata without the "?"